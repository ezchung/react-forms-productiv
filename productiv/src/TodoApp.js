import React, { useState } from "react";
import { v4 as uuid } from "uuid";

import TopTodo from "./TopTodo";
import EditableTodoList from "./EditableTodoList";
import TodoForm from "./TodoForm";


/** App for managing a todo list.
 *
 * Props:
 * - initialTodos: possible array of [ todo, ... ]
 *
 * State:
 * - todos: array of [ todo, ... ]
 *
 * App -> TodoApp -> { TodoForm, EditableTodoList }
 */

function TodoApp({initialTodos}) {
  const [todos, setTodos] = useState(initialTodos);

  const initialFormInfo = {
    title: "",
    description: "",
    priority: 2
  } //TODO: Can move this todo app. Assign to default

  console.log("TODOAPP todos ----->", todos)

  /** add a new todo to list */ 
  function create(newTodoInfo) {
    //Takes in information
    let id = uuid();
    let newTodo = {...newTodoInfo, id: id}; //TODO: id: id ==> , id
    setTodos(todo => [...todo, newTodo]);   //Give todo name todos. Not singular
  }

  /** update a todo with updatedTodoInfo */ 
  function update(updatedTodoInfo) {
    let updateTodo = {...updatedTodoInfo};
    let updatingTodos = todos.filter(todo => todo.id !== updateTodo.id); 
    setTodos(todo => [...updatingTodos, updateTodo]); //Rename old state to plural
  }

  /** delete a todo by id */ 
  function remove(id) {
    let filterTodos = todos.filter(todo => todo.id !== id); 
    setTodos(todo => [...filterTodos]); //Rename old state to plural
  }


  return (
      <main className="TodoApp">
        <div className="row">


          <div className="col-md-6">
            {todos.length < 1
              ? <span className="text-muted">You have no todos.</span>
              : <><h1>Todos</h1>
              <EditableTodoList
                  todos={todos}
                  update={update}
                  remove={remove}/>  </>}
          </div>

          <div className="col-md-6">
            <section className="mb-4">
              <h3>Top Todo</h3>
              <TopTodo todos={todos}/>
            </section>

            <section>
              <h3 className="mb-3">Add NÃ¼</h3>
              <TodoForm initialFormData={initialFormInfo} handleSave={create}/>
            </section>
          </div>

        </div>
      </main>
  );
}

export default TodoApp;