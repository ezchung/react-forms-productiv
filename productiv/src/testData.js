import React from "react";
import { render } from "@testing-library/react";

const TODO_1 = {
  id: 1,
  title: "test1",
  description: "test description1",
  priority: 1
}

const TODO_2 = {
  id: 2,
  title: "test2",
  description: "test description2",
  priority: 2
}

const TODO_3 = {
  id: 3,
  title: "test3",
  description: "test description3",
  priority: 3
}

const TODO_4 = {
  id: 4,
  title: "test4",
  description: "test description4",
  priority: 2
}

const TODO_5 = {
  id: 5,
  title: "test5",
  description: "test description5",
  priority: 3
}


export {TODO_1, TODO_2, TODO_3, TODO_4, TODO_5};