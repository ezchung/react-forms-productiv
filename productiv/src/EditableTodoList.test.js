import React from "react";
import { render } from "@testing-library/react";
import EditableTodoList from "./EditableTodoList";
import { TODO_4, TODO_5} from './testData';
import TodoApp from "./TodoApp";


TodoApp.update = jest.fn();
TodoApp.remove = jest.fn();


/*********** Render and Smoke Tests ****************/

it("renders without crashing", function() {
  render(<EditableTodoList todos={[TODO_4]} update={TodoApp.update} remove={TodoApp.remove} />);
})

it("matches snapshot", function() {
  const {container} = render(<EditableTodoList todos={[TODO_5]} />);
  expect(container).toMatchSnapshot();
});

/*********** TEST DISPLAY ****************/
it("displays all todos", function () {
    const {container } = 
        render(<EditableTodoList 
                    todos={[TODO_4, TODO_5]} 
                    update={TodoApp.update} 
                    remove={TodoApp.remove}
                />
    );

    expect(container).toContainHTML("test description5");
    expect(container).toContainHTML("test description4");
    expect(container).not.toContainHTML("test description1");
});
