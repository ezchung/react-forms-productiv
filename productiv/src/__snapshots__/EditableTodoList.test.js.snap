// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <div
      class="EditableToDo"
    >
      <div
        class="mb-3"
      >
        <div
          class="float-end text-sm-end"
        >
          <button
            class="EditableTodo-toggle btn-link btn btn-sm"
          >
            Edit
          </button>
          <button
            class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
          >
            Del
          </button>
        </div>
        <div>
          <div
            class="Todo"
          >
            <div>
              <b>
                test5
              </b>
               
              <small>
                (priority: 
                3
                )
              </small>
            </div>
            <div>
              <small>
                test description5
              </small>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
